CodeReady Workspace quick guide

j-c-berger opened the pull request:
For issue [#2407-Support for CodeReady Workspaces](https://github.com/eclipse/codewind/issues/2407) 

@AndyRWatson, tagging you since you opened the issue. @micgibso, tagging you for your awareness as well. 

I'm not sure if these instructions for developing a microservice are applicable for CodeReady Workspaces (I took these instructions from our other quick guides), but I thought it at least gives us a starting point. 

Note, I took a lot of this content for [Knowledge Center's Installing CodeReady Workspaces](https://www.ibm.com/support/knowledgecenter/SSCSJL_4.1.x/docs/ref/general/installation/installing-codeready-and-codewind.html). I worked on this content a couple months ago.

j-c-berger commented on docs/_guides/codewind-crw-quick-guide.md:
```diff
@@ -0,0 +1,196 @@
+Each Codewind workspace also requires at least on 1Gi ReadWriteMany (RWX) persistent volume.
+
+### Installing CodeReady Workspaces
+To install CodeReady Workspaces, set `Spec.codeReadyWorkspaces.enable: true` in the Kabanero custom resource (Kabanero CR) instance and apply it.
```
I'll changes these specs to highlight the proper text. I'll wait to make the change once we have the new instructions (the instructions without Kabanero references).

j-c-berger commented on docs/_guides/codewind-crw-quick-guide.md:
```diff
@@ -0,0 +1,196 @@
+
+To edit the Kabanero CR, you can run `oc edit kabanero -n kabanero` from the command line. 
+
+The following sample shows a Kabanero CR instance configuration: 
```
Same as my previous comment.

j-c-berger commented:
@johnmcollier, new commit is up. If it looks good, I can get SVT to run through and test it.

johnmcollier commented on docs/_guides/codewind-crw-quick-guide.md:
```diff
@@ -0,0 +1,206 @@
+### Prerequisite
+CodeReady Workspaces requires at least two 1Gi ReadWriteOnce (RWO) persistent volumes on the cluster to install and a 1Gi RWO volume for each created workspace.
+
+Each Codewind workspace also requires at least on 1Gi ReadWriteMany (RWX) persistent volume.
```
This should be updated to 5Gi

johnmcollier commented on docs/_guides/codewind-crw-quick-guide.md:
```diff
@@ -0,0 +1,206 @@
+
+### Installing CodeReady Workspaces
+1. Subscribe to the CodeReady Workspaces operator then install it.
+  * To manually install CodeReady Workspaces, apply the CodeReady Workspaces OperatorHub subscription. 
```
Maybe we can take this sub point out? Let me also get you some screenshots for these steps, as I actually do think it will help now.

johnmcollier commented on docs/_guides/codewind-crw-quick-guide.md:
```diff
@@ -0,0 +1,206 @@
+### Installing CodeReady Workspaces
+1. Subscribe to the CodeReady Workspaces operator then install it.
+  * To manually install CodeReady Workspaces, apply the CodeReady Workspaces OperatorHub subscription. 
+2. After the CodeReady Workspaces operator is installed, create the CheCluster resources. 
```
This step can be removed, since we're just focusing on Codewind at the moment

johnmcollier commented on docs/_guides/codewind-crw-quick-guide.md:
```diff
@@ -0,0 +1,206 @@
+  * If you have not set up an auth provider, set `openShiftoAuth` to `false`. 
+  * To enable HTTPS, set `tlsSupport` to `true`.
+  * If you use a self-signed certificate, set `selfSignedCert` to `true`. 
+3. If you install Codewind alongside CodeReady Workspaces, use the following yaml:  
```
Change to something like "use the following yaml"? Again I think screenshots may help in this step, so I'll get you some tomorrow.

johnmcollier commented on docs/_guides/codewind-crw-quick-guide.md:
```diff
@@ -0,0 +1,206 @@
+### Setting up Codewind
+Before you can use Codewind with CodeReady Workspaces or Che, complete the following steps, either before or after you install Codewind: 
+
+1. Apply the Codewind cluster role with `oc apply -f`. 
```
I think this should be a prerequisite step before installing CRW

j-c-berger replied:
On this note, do we need to include additional instructions to install Codewind? Like a link to our existing instructions to install Codewind?

j-c-berger commented:
@johnmcollier, thanks, I addressed your comments and made a new commit. I reordered some content, per your comments. I'll include the screenshots once you get them to me. 

@naveenkaratekid, I think it's ready for you to test it since we're only waiting on screenshots now. Thanks!

j-c-berger commented:
@johnmcollier and @naveenkaratekid, new commit with images included is up, thanks!

j-c-berger commented on docs/_guides/codewind-crw-quick-guide.md:
```diff
@@ -0,0 +1,213 @@
+Before you can use Codewind with CodeReady Workspaces, apply the Codewind cluster role with `oc apply -f`. 
+
+### Installing CodeReady Workspaces
+1\. Subscribe to the CodeReady Workspaces OperatorHub then install CodeReady Workspaces:
```
@johnmcollier, @naveenkaratekid was going through the guide, and we were both stuck at this first step. How do we subscribe to CRW OperatorHub then install CRW? From an online search, I did not find any CRW OperatorHub.

naveenkaratekid replied:
I also wasn't able to apply the yaml file in step 2. I get this error: 
```
Naveens-MacBook-Pro:codeready naveenk@ibm.com$ oc apply -f file.yaml
error: unable to recognize "file.yaml": no matches for kind "CheCluster" in version "org.eclipse.che/v1"
```

johnmcollier replied:
The steps here aren't quite accurate:
1. Log on to your OpenShift cluster
2. From the sidebar, click on `Operators` then `OperatorHub`
3. Search for `CodeReady Workspaces`
4. Select it and click install

j-c-berger replied:
I'll get these new instructions in shortly.

j-c-berger commented on docs/_guides/codewind-crw-quick-guide.md:
```diff
@@ -0,0 +1,213 @@
+  * To enable HTTPS, set `tlsSupport` to `true`.
+  * If you use a self-signed certificate, set `selfSignedCert` to `true`. 
+
+2\. Use the following yaml:  
```
Another good point Naveen raised, for this yaml file, do we save the file and use `oc apply -f <file>`?

johnmcollier replied:
To install CRW (once its operator has been installed):
1) From the OpenShift dashboard, go to `Operators` -> `Installed Operators`
2) Select `CodeReady Workspaces`
3) Select `CodeReady Workspaces` cluster. Then `Create CheCluster`
4) Copy and paste the yaml into the text box you see. And configure the fields listed above (such as `tlsSupport` as needed.

naveenkaratekid replied:
@johnmcollier I don't see the CodeReady workspace in my console. Is there a way to install it?

johnmcollier replied:
@naveenkaratekid What version of OCP are you testing against? 4.x or higher will be needed

naveenkaratekid replied:
@johnmcollier I believe I'm running OCP 4.3 from FYRE

johnmcollier replied:
@naveenkaratekid Have you tried the steps in https://github.com/eclipse/codewind-docs/pull/613#discussion_r436721067?

naveenkaratekid replied:
@johnmcollier I just ran these two commands:
```
oc adm policy add-scc-to-user anyuid system:serviceaccounts:default:che-workspace
oc adm policy add-scc-to-user privileged system:serviceaccounts:default:che-workspace
```
The docs don't clearly explain how to start codewind using codeready. The guide says 
```
CodeReady Workspaces starts Codewind and installs the Codewind plug-ins. This process might take a couple of minutes for all of the necessary components to be pulled and started.

But after I refreshed the console, and clicked Resources, I didn't see Codewind show up there
```

johnmcollier replied:
@naveenkaratekid Once you've installed CodeReady Workspaces, you need to log in and create a Codewind workspace from the Codewind devfile, as you would for Che (basically https://www.eclipse.org/codewind/che-createcodewindworkspace.html)

@j-c-berger Maybe we need to add the steps from https://www.eclipse.org/codewind/che-createcodewindworkspace.html in to this guide?

j-c-berger replied:
@johnmcollier, I'm happy to include these steps. I'm thinking we should include these new steps before we say, `CodeReady Workspaces now installs and you can access the Codewind Workspace URL and Red Hat SSO Admin Console URL`. So putting these new steps after the yaml file. 

Agree?

johnmcollier commented on docs/_guides/codewind-crw-quick-guide.md:
````diff
@@ -0,0 +1,213 @@
+
+2\. Use the following yaml:  
+```yaml
+apiVersion: org.eclipse.che/v1
````
This yaml needs to be updated to:
```
apiVersion: org.eclipse.che/v1
kind: CheCluster
metadata:
  name: codeready-workspaces
spec:
  server:
    cheImageTag: ''
    cheFlavor: codeready
    devfileRegistryImage: ''
    pluginRegistryImage: ''
    tlsSupport: true
    selfSignedCert: true
    cheWorkspaceClusterRole: 'eclipse-codewind'
    customCheProperties:
      CHE_INFRA_KUBERNETES_WORKSPACE__START__TIMEOUT__MIN: "15"
      CHE_LIMITS_WORKSPACE_IDLE_TIMEOUT: "0"
      CHE_WORKSPACE_PLUGIN__BROKER_WAIT__TIMEOUT__MIN: "15"
  database:
    externalDb: false
    chePostgresHostName: ''
    chePostgresPort: ''
    chePostgresUser: ''
    chePostgresPassword: ''
    chePostgresDb: ''
  auth:
    openShiftoAuth: false
    identityProviderImage: ''
    externalIdentityProvider: false
    identityProviderURL: ''
    identityProviderRealm: ''
    identityProviderClientId: ''
  storage:
    pvcStrategy: per-workspace
    pvcClaimSize: 1Gi
    preCreateSubPaths: true
```

j-c-berger replied:
On it. Will have the new commit up soon.

j-c-berger commented:
@johnmcollier and @naveenkaratekid, new commit is up which includes the new instructions and yaml file John provided. 

Should we include instructions or a link to setting up an OpenShift cluster? The first step for `Installing CodeReady Workspaces` requires users to log into their OpenShift clusters, and I wonder if some users won't have an OpenShift cluster already. 

I have the same thought when we say in the Prerequisites to `apply the Codewind cluster role with oc apply -f`. I thought I could provide more context to users. Are they applying the `oc apply -f` command in the command line? 

I think my images are showing up as broken links because they don't exist in the `images` folder. They only exist on this PR. So let me know if you have any questions about the order of the screenshots. 

Thanks!

j-c-berger commented:
@johnmcollier, please see new commit and let me know if the instructions are accurate and what you intended. Same goes for placement of screenshots. 

@naveenkaratekid, if you could test it today with these new instructions, that'd be great!

johnmcollier commented:
@j-c-berger I think we just need to make an OpenShift cluster a prerequisite for the quick guide. 

And yes, the `oc apply` commands are done on the command line.

johnmcollier commented on docs/_guides/codewind-crw-quick-guide.md:
```diff
@@ -0,0 +1,244 @@
+
+![Subscribe to OperatorHub](./images/codereadyworkspaces-images/CRW-OperatorHub.png){:width="900"}. 
+
+![Install CodeReady Workspaces](./images/codereadyworkspaces-images/CRW-Install.png){:width="800"}. 
```
This, the screenshot, and the optional configurations should be below the sample yaml

johnmcollier commented on docs/_guides/codewind-crw-quick-guide.md:
```diff
@@ -0,0 +1,244 @@
+
+After you install CodeReady Workspaces, log in and create a Codewind workspace from the Codewind devfile: 
+
+1\. Log in to Che. Che loads. 
```
Change Che to CodeReady Workspaces

johnmcollier commented on docs/_guides/codewind-crw-quick-guide.md:
```diff
@@ -0,0 +1,244 @@
+
+4\. From **Source**, click **YAML**. 
+
+5\. Go to the link, [codewind-che-plugin/0.12.0/devfile.yaml](https://raw.githubusercontent.com/eclipse/codewind-che-plugin/0.12.0/devfiles/0.12.0/devfile.yaml), then copy and paste the contents into the YAML text box in your Codewind workspace. 
```
Let's use 0.13.0 instead: https://raw.githubusercontent.com/eclipse/codewind-che-plugin/0.13.0/devfiles/0.13.0/devfile.yaml

johnmcollier commented on docs/_guides/codewind-crw-quick-guide.md:
```diff
@@ -0,0 +1,244 @@
+
+### Setting up Codewind
+Because of its dependency on `buildah`, Codewind needs to run as root and privileged. To enable Codewind, run the following commands: 
+  * `oc adm policy add-scc-to-user anyuid system:serviceaccounts:<namespace>:che-workspace` 
```
We should say "where <namespace> is the namespace that you installed CodeReady Workspaces in to"

johnmcollier commented on docs/_guides/codewind-crw-quick-guide.md:
```diff
@@ -0,0 +1,244 @@
+### Configuring Codewind to use application stacks
+Configure Codewind to use Appsody templates so you can focus exclusively on your code. Complete the following steps to select the Appsody templates:
+
+1. Under the Explorer pane, select **Codewind**.
```
These instructions are not quite right. It's still specific to Local Codewind and not remote.

1. On the right hand side, select the Codewind logo
2. Right click on `Projects`
3. Select **Template Source Manager**.
4. Enable **Appsody Stacks - incubator** and **Default templates**.

johnmcollier commented on docs/_guides/codewind-crw-quick-guide.md:
```diff
@@ -0,0 +1,244 @@
+
+After you configured Codewind to use Appsody templates, continue to develop your microservice within Codewind.
+
+If your organization uses customized application stacks and gives you a URL that points to an `index.json` file, you can add it to Codewind:
```
We can remove this

johnmcollier commented on docs/_guides/codewind-crw-quick-guide.md:
```diff
@@ -0,0 +1,244 @@
+### Working with the example calculator microservice
+You now can work with the example calculator microservice.
+
+1. Use the port number that you saw when you first opened the application.
```
Change this to `Use the URl that you saw when you first opened the application`

johnmcollier commented on docs/_guides/codewind-crw-quick-guide.md:
```diff
@@ -0,0 +1,244 @@
+
+1. Use the port number that you saw when you first opened the application.
+2. Make sure to remove the `< >` symbol in the URL.
+3. `http://127.0.0.1:<port>/starter/calculator/aboutme`
```
Change this to `<url>/starter/calculator/aboutme`

johnmcollier commented on docs/_guides/codewind-crw-quick-guide.md:
```diff
@@ -0,0 +1,244 @@
+
+You can also try a few of the sample calculator functions:
+
+* `http://127.0.0.1:<port>/starter/calculator/{op}/{a}/{b}`, where you can input one of the available operations `(+, _, *)`, and an integer a, and an integer b.
```
`http://127.0.0.1:<port>` to `<url>`

j-c-berger commented:
@johnmcollier, thank for the review! New commit is up, reflecting the changes. 

For the prerequisites, I know we want to add installing an OpenShift cluster. I know we talk about clusters some in our [installing Eclipse Che doc](https://www.eclipse.org/codewind/che-installinfo.html). I didn't know if you wanted to adapt some of the instructions there.

naveenkaratekid commented:
@johnmcollier Step 6 in the guide is not very clear
```
CodeReady Workspaces now installs and you can access the CodeReady Workspaces URL and Red Hat SSO Admin Console URL
```

I assumed that I would need to click on this link ```codeready-default.apps.cretic.os.fyre.ibm.com ```, but when I did that, it took me to this page.

![image](https://user-images.githubusercontent.com/19738071/84193239-1c38e480-aa69-11ea-80f5-bd923470af6d.png)

The guide didn't state how we are supposed to log in

johnmcollier commented:
@naveenkaratekid You'll need to click register to create an account.

@j-c-berger Can you add that to the post-install instructions? Basically before we log in to CodeReady Workspaces to create a codewind workspace

AndyRWatson commented:
I just tried following the guide, using the updates.  When running prerequisite steps,  I setup a OCP on IBM cloud,   then I assumed I should login from the command  (need to state step) and the performed 

```oc apply -f ```

and got the response 

```error: error reading ".": open .Trash: operation not permitted```

I assume more is required to apply a codewind cluster or there maybe a missing step(s) here ?

AndyRWatson commented:
For the step 
```Installing CodeReady Workspaces

1. Log in to your OpenShift cluster.

```

is this still command line or webui ?     I assume this is the Openshift web console ?

AndyRWatson commented:
```
    If you did not set up an auth provider, set openShiftoAuth to false.
    To enable HTTPS, set tlsSupport to true.
    If you use a self-signed certificate, set selfSignedCert to true.
```
For a guide I think we should let people know how to do this,   I think  some people may drop out here unless we are prerequing users must be skilled at OpenShift setup and configuration ?

AndyRWatson commented:
when installing the codeready operator,  is the recomendation step 4 or step 5 ?   I could update my tls in the editor in 4 and then overwrite that change via c+p in step 5 ?

AndyRWatson commented:
```Install CodeReady Workspaces{:width="800"}.

    To manually install Codewind with CodeReady Workspaces, save this yaml to a disk and run oc apply -f <yaml>.
    If you did not set up an auth provider, set openShiftoAuth to false.
    To enable HTTPS, set tlsSupport to true.
    If you use a self-signed certificate, set selfSignedCert to true.
```

how do I save the yaml to a disk ?  And if we're now going back to the command line window maybe we should clarify the switch    For the next steps do we set auth/https/certs via the command line, in the yaml or by going back to the openshift console ?

AndyRWatson commented:
```6. CodeReady Workspaces now installs and you can access the CodeReady Workspaces URL and Red Hat SSO Admin Console URL:```

How do I discover the URLs ?

micgibso commented on docs/_guides/codewind-crw-quick-guide.md:
```diff
@@ -0,0 +1,243 @@
+duration: 15 minutes
+tags: Codewind, CodeReady Workspaces, Openshift, Kubernetes, containers, microservice
+--- 
+
```
Hi Jake, just spotted that the metadata needs to be tweaked. The objectives are now part of the metadata, here's an example from the latest VS Code Guide:
```
---
layout: guide
summary_title: "Codewind in VS Code"
title: "Getting Started with Codewind in VS Code"
categories: guides
description: "Take advantage of Codewind's tools to help build high quality cloud native applications regardless of which IDE or language you use."
permalink: codewind-vscode-quick-guide.html
duration: 5 minutes
`keywords: Codewind, VS Code, microservice`
`objectives: ["Install Visual Studio Code (VS Code) and Codewind.", "Develop a simple microservice that uses Eclipse Codewind in VS Code."]`
`icon: images/learn/icon_logoVScode.svg`
---
```
So change `guide_description:` to `description:`
And add:
```
keywords: Codewind, CodeReady, Workspace
objectives: ["Install CodeReady Workspaces and Codewind.", "Develop a simple microservice that uses Eclipse Codewind on CodeReady Workspaces."]
icon: images/learn/icon_logoVScode.svg    
```

For the `objectives:` keep the syntax exactly as I've typed above including the `[` and `]` .
For the icon image, we need a Codeready Workspaces icon.

micgibso commented on docs/_guides/codewind-crw-quick-guide.md:
```diff
@@ -0,0 +1,243 @@
+## Objectives
+* Install CodeReady Workspaces and Codewind
+* Develop a simple microservice that uses Eclipse Codewind on CodeReady Workspaces
+
```
Delete the Objectives (lines 12, 13, and 14) from the content because they should be in the metadata; see previous comment.

j-c-berger commented:
@johnmcollier, please see new commit, and let me know if the content looks look. Per the issues with Appsody on Kube, the guide focuses solely on installing CodeReady Workspaces and Codewind and setting up a Codewind workspace. 

@micgibso, as you'll see, I made the metadata changes. However, this is what I see when I render the guide locally:   
<img width="1332" alt="Screen Shot 2020-06-11 at 12 06 42 PM" src="https://user-images.githubusercontent.com/54117431/84416755-a193d480-abe2-11ea-8705-9e169f6b27fb.png">
I believe it's because my branch was made off master before James made these changes to guides' metadata. I think once this PR is merged, the objectives and icon will display properly because the branch will then be in sync with the newest master. I wanted to confirm with you though before anything is  merged. Thanks!

johnmcollier commented on docs/_guides/codewind-crw-quick-guide.md:
```diff
@@ -0,0 +1,136 @@
+CodeReady Workspaces uses Kubernetes and containers to provide a preconfigured environment. Use CodeReady Workspaces to create, build, and test your code in OpenShift containers but feel like you are working on an IDE on your local machine.  
+
+### Prerequisite
+CodeReady Workspaces require at least two 5Gi ReadWriteOnce (RWO) persistent volumes on the cluster to install and a 5Gi RWO volume for each created workspace.
```
We should highlight that OpenShift itself is prerequisite for this lab. And maybe recommend "CodeReady Containers" (https://code-ready.github.io/crc/) if they don't have OpenShift already available.

j-c-berger replied:
Thanks, John. I'll update the prereqs this afternoon. If all else looks good to you, I can see if Michael can merge tomorrow morning.

johnmcollier replied:
Yup, everything else looks good

j-c-berger commented:
@micgibso, PR is ready to merge. @johnmcollier reviewed it and had one change to the prerequisite. See the new commit for those changes. 

Also, please note my earlier comment about the metadata and this branch being made off an old form of master. I don't think that will be an issue after merging this PR but thought you should know. Thanks!
